### DISCUSSION TITLE
TypeError: Cannot read property 'uid' of null in form and edit form

### DISCUSSION THREAD

**Discussion by cmeffley** [nss-evening-cohort-14/evening-client-side#304] [Capstones!] [answered]
Chris Meffley

Similar to my other ticket, but

Now getting TypeError: Cannot read property 'athleteUid' of null when trying to add or edit race, and names not appearing in dropdown menu. Also race cards no longer show.
I believe this is all related.

Searched:
https://stackoverflow.com/questions/52254513/typeerror-cannot-read-property-uid-of-null/52292849
https://stackoverflow.com/questions/41446560/react-setstate-not-updating-state
https://stackoverflow.com/questions/51533945/typeerror-cannot-read-property-of-null-in-react-component

Tried:
- Added isMounted in race view component to make sure all the promise calls came back before the page was loaded
- Trying to create a separate useState hook for coach races in race form view and then do a ternary for which races are mapped and showing (it does not appear to like the ternary around the map function).

Files:
https://github.com/cmeffley/front-end-capstone/blob/crudraces/src/App/index.js
https://github.com/cmeffley/front-end-capstone/blob/crudraces/src/views/RacesFormView.js
https://github.com/cmeffley/front-end-capstone/blob/crudraces/src/components/RacesForm.js
https://github.com/cmeffley/front-end-capstone/blob/crudraces/src/components/RaceSchedule.js
https://github.com/cmeffley/front-end-capstone/blob/crudraces/src/views/RacesFormView.js
https://github.com/cmeffley/front-end-capstone/blob/crudraces/src/helpers/data/athleteData.js

**Comment by drteresavasquez** [comment]
@cmeffley if you do not already have a flow chart for your data flow, I would start with that. I think you have had a similar problem and I want you to work through this one. Check your logic and flow and ensure that your API calls are returning what you expect

**Reply by cmeffley** [reply]
To rubber duck my work-through:
My state is not updating, because my get races API call is not working. When I do the API call like this:
```axios.get(`${dbUrl}/races.json?orderBy="coachUid"&equalTo="hard-coded coachUid"`)``` the array shows up (but the null error still persists), but when I put in my taco in the `&equalTo-"${taco}"` the array is empty.

**Reply by drteresavasquez** [reply]
Are you setting state to null initially? If so, try setting it to an array instead

**Reply by cmeffley** [reply]
It is set to array initially: ```const [races, setRaces] = useState([]);```

**Reply by drteresavasquez** [reply]
Make sure your rules allow you to index on coachUid

**Reply by cmeffley** [reply]
Yes, right now, I have indexOn for my races, athlete and coach tables.

**Comment by TrinityChristiana** [comment, marked as answer]
@cmeffley Will add the solution!

**Reply by cmeffley** [reply]
1. My get races API call was returning an empty array, because it was looking for a specific user logged in, in an array of users (i.e. coach array has one object in it), so therefore, when making my API call for which user was logged in, I needed to specify the index of the user, so the races API knew where to look for the races.
2. Once the races and athlete (user) was known, my form stated that the athlete uid was undefined, because in my state, I had it written out specifically (from the prop), when I needed to just get it from the race info object:
``` athleteUid: athlete.athleteUid || '' ``` <----incorrect
``` athleteUid: raceInfo?.athleteUid || '' ``` <----correct
3. In my handle submit function, I was looking at the athlete and coach uid's (from props) in order to edit and add races. I needed to replace them, with raceInfo (`athlete.athleteUid` changed to `raceInfo.athleteUid`) because, again, I needed to look specifically at the info object that was already there.
4. In my drop down menus where I can select a coach or athlete, the name displayed needed to come from the function parameter, not the athlete or coach prop.
